### ISSUE TITLE
Add wheel builds

### ISSUE THREAD

**Issue by scaramallion** [pydicom/pylibjpeg-libjpeg#3]
It may be annoying, but at least *try* and get it working on windows...

Will need to provide wheels for 32 and 64-bit windows for Python 3.6, 3.7 and 3.8, which will be interesting. How do other projects do this?

Appveyor has a large range of WIndows VMs...

https://github.com/joerick/cibuildwheel and https://github.com/pypa/manylinux

OK I think I love cibuildwheel :heart_eyes: 

---

Hah, I can't get an API token from pypi without a project and I can't upload a whl file to pypi to create a project because I don't have an API token.

**Comment by scaramallion** [comment]
Wheels are being built on Linux, OSX and Windows and uploading to test PyPi OK. Need to test them to make sure they work (particularly Windows).

**Comment by scaramallion** [comment]
Windows wheels working on non-dev machine! 

Tested: 
32-bit - 3.8

Untested: 
32-bit - 3.8
64-bit - 3.6, 3.7, 3.8

**Comment by scaramallion** [comment]
Fresh Windows 10 in a VM, has `msvcp140_clr0400.dll` but not `MSVCP140.dll`. 32- and 64-bit work with Python 3.6, 3.7 and 3.8.

**Comment by scaramallion** [comment]
@mrbean-bremen @jwitos do you mind testing out the wheels for `pylibjpeg-libjpeg`?

### Windows
Wheels: https://transfer.sh/11aZlZ/wheelhouse-win.tar.gz
sha256: `6865b2b22d7c2ef501c3efc1c4c2822ad89166fac2779cc7f9f6073960940019`

### OSX
Wheels: https://transfer.sh/jnotG/wheelhouse-osx.tar.gz
sha256: `5cbfd635a6040e37cdcb213a550ccdf4fb2018aa18cde77d71507512e2d2afa1`

### Linux
Wheels: https://transfer.sh/wkRb3/wheelhouse-linux.tar.gz
sha256: `76514a599530f3d2bab03511a71917bfed779a96c9eda9ee99971fbe26e6745f`

Download, check sha256 hash against tar file, extract and install the corresponding wheel to your setup with `pip install path/to/*.whl`. It won't work with pydicom straight away, you need to add a pixel data handler:
```python
from pydicom import dcmread
from pydicom.data import get_testdata_file
import libjpeg
from libjpeg.tests import add_handler
import matplotlib.pyplot as plt

add_handler()

ds = dcmread(get_testdata_file("JPEG-LL.dcm"))
print(ds.file_meta.TransferSyntaxUID)
arr = ds.pixel_array

plt.imshow(arr)
plt.show()
```

If you could post your OS, 32/64 bit arch, Python version and any errors you get, that'd be a huge help.

**Comment by mrbean-bremen** [comment]
I may get to it somewhere in the next days, have been travelling the last week. Awesome work, anyway!

**Comment by scaramallion** [comment]
Whenever you have time is fine. @SimonBiggs you might be interested in this as well?

**Comment by SimonBiggs** [comment]
Yes certainly! Thank @scaramallion. This is way tidier than my hacky solution:

https://github.com/pymedphys/pymedphys/blob/master/pymedphys/_losslessjpeg/__init__.py

Great to see a Python wrapped, Windows supported losslessjpeg library in the wild :).

**Comment by mrbean-bremen** [comment]
Ok, did some tests under Windows (64 bit):

- Download, verify checksum, extract wheels
- Install in Python 3.6.3 venv, uninstall Pillow, install pylibjpeg-data
- Create test file, run it -> 
```
Traceback (most recent call last):
  File "jpeg_plugin.py", line 3, in <module>
    import libjpeg
  File "c:\dev\venv\py3_test\lib\site-packages\libjpeg\__init__.py", line 6, in <module>
    from .utils import decode, get_parameters
  File "c:\dev\venv\py3_test\lib\site-packages\libjpeg\utils.py", line 8, in <module>
    import _libjpeg
  File "__init__.pxd", line 918, in init _libjpeg
ValueError: numpy.ufunc size changed, may indicate binary incompatibility. Expected 216 from C header, got 192 from PyObject
```
numpy version: 1.14.2 
- pip install -U numpy -> numpy 1.18.1
- run test file -> all fine, image looks good
- tested other numpy versions -> works with numpy >= 1.16.0

- same procedure with Python 3.7.1 env -> works nicely
- same with Python 3.8.1 env -> numpy was not installed, latest numpy was installed by wheel
- installed matplotlib, run test -> works fine

**Comment by mrbean-bremen** [comment]
I'm not sure where to run pytest, though - shall I run the pydicom tests with this plugin?

**Comment by scaramallion** [comment]
Oh yeah, I should've specified how to run pytest: `python -c "import pytest; pytest.main(['--pyargs', 'libjpeg.tests'])"`. Running with pydicom tests more of the decoding and would be helpful.

So as I understand it, reinstalling the higher version numpy fixed the ufunc exception? I wonder if it will reoccur if I force 1.14.2...

**Comment by scaramallion** [comment]
Related: https://github.com/numpy/numpy/issues/12785

**Comment by mrbean-bremen** [comment]
Thanks, will test this tomorrow.
And yes, the error was gone with NumPy >= 1.16.0.

**Comment by mrbean-bremen** [comment]
This is for the 3.7 wheel:
```
(pydicom-3.7.1) c:\dev\GitHub>python -c "import pytest; import libjpeg; pytest.main(['--pyargs', 'libjpeg.tests'])"
============================= test session starts =============================
platform win32 -- Python 3.7.1, pytest-3.9.3, py-1.7.0, pluggy-0.8.0
rootdir: c:\dev\GitHub, inifile:
collected 124 items

tests\test_decode.py ...........................................         [ 34%]
tests\test_environment.py sss                                            [ 37%]
tests\test_handler.py .....ss....s.....ss..s................             [ 67%]
tests\test_parameters.py ..s...................................          [ 98%]
tests\test_reconstruct.py .s                                             [100%]

=================== 113 passed, 11 skipped in 70.13 seconds ===================
```
Exactly the same result for 3.6 and 3.8, so all looks good!

Note: I used `python -c "import pytest; import libjpeg; pytest.main(['--pyargs', 'libjpeg.tests'])`, e.g. added `import libjpeg` (for anyone else who wants to run the tests).

**Comment by scaramallion** [comment]
Awesome, thanks!

**Comment by jwitos** [comment]
@scaramallion I just wanted to apologize for not responding -- last week was very busy and I did not have a chance to test the wheels. Looks like they're not available anymore and

I just checked the packaged `pylibjpeg-libjpeg`, though. It worked perfectly for both default pydicom lossless data, as well as my own dataset with `1.2.840.10008.1.2.4.70` Syntax UID and regular non-lossless data. I tested it on Windows 10 64bit with conda env, Python 3.8 and numpy 1.18.1.

Sorry again!
